### PR TITLE
Fixed an issue where the wrong page title was returned

### LINKED ISSUE #6308: Page.get_title_obj doesn't return correct title under certain conditions
### Summary

Calling `Page.get_title_obj` from a template on a page that has a title in french (configured to fallback to english) and a title in english, while viewing the page in french, will return the english title. This issue only happens if the page has an associated extension object (for example when using [this example code](http://docs.django-cms.org/en/latest/how_to/extending_page_title.html#title-model-extension-example)). 

The bug doesn't occur with DjangoCMS 3.4.5, and the issue seems to have been introduced by the removal of this line https://github.com/divio/django-cms/commit/f7af69f3e1#diff-960fbbc1702d77aa625a17bcb0c37191L1281. In my case the title cache only contains the french version of the page for as long as the page doesn't have any extension. As soon as I add an extension object, the cache then only contains the english version of the page, causing the call to `Page.get_title_obj` to use the english version of the page.

### Expected behaviour

`Page.get_title_obj` returns the page title in the current language.

### Actual behaviour

`Page.get_title_obj` returns the fallback language.

### Environment

* Python version: 3.4.2
* Django version: 1.11.11
* django CMS version: 3.5.1

### PR DESCRIPTION
**Pull request by mekanics**
### Summary

Fixed an issue where a page with a TitleExtension always returned the fallback title. 

Force to reload the title cache if the language was not found,
before it returns the appropriate language.
The issue occured, when a page with multi languages had a TitleExtension.

Fixes #6308

### Proposed changes in this pull request
Reordering the code inside the `__get_title_cache` method to first checking if 
a 'force_reload' is needed and then returning the appropriate language. 
Additionally we can remove some duplicated code. Makes the method smaller and easier to understand.

### Documentation checklist

* [x] I have updated CHANGELOG.txt if appropriate
* [x] I have updated the release notes document if appropriate, with:
    * [ ] general notes
    * [x] bug-fixes
    * [ ] improvements/new features
    * [ ] backwards-incompatible changes
    * [ ] required upgrade steps
    * [ ] names of contributors
* [ ] I have updated other documentation
* [x] I have added my name to the AUTHORS file
* [ ] This PR's documentation has been approved by Daniele Procida

### CODE REVIEW

**Comment by madEng84** [comment]
Hi @mekanics 
I have wrote some test for the problem in this [issue request](https://github.com/divio/django-cms/issues/6308/#issuecomment-429425359)

The first two items of my [tests](https://github.com/divio/django-cms/files/2474102/tests.py.txt) maybe have an unexpected behaviour, as you can read in the issue request.
Can you tell me if they are ok with your patch?
If so and if you want you can include them in your commit.

Thank you, 
if I can help you I'm here.

**Comment by mekanics** [comment]
Hi @madEng84 

I did try your tests today and I'm having the same behaviour as you. 

both failing tests do not redirect to the 'en' page, but they do return an HTML File with: 
```
This page has no preview!
It is being redirected to: /en/contacts/work-with-us/
```
which IMO is probably the correct behaviour as we are in edit mode **on** (?)

However I'm getting a redirect (302) if I make the requests in edit mode **off**.

**Comment by madEng84** [comment]
Thank you so much @mekanics 

You're right!
The problem is that i was not able to reproduce the problem with a test using a staff user. (maybe because he hasn't enough permissions)

You saw that message ("This page has no preview!") because staff user see an emulated 302 response as you can see in these django-cms statements
 
_cms/views.py_
```
def details(request, slug):
    ...

    if redirect_url:
        if request.user.is_staff and toolbar.edit_mode_active:
            toolbar.redirect_url = redirect_url
        elif redirect_url not in own_urls:
            # prevent redirect to self
            return HttpResponseRedirect(redirect_url)
```

_templates/cms/toolbar/toolbar_with_structure.html_
```
{% if cms_toolbar.redirect_url %}
        <div class="cms-screenblock">
            <div class="cms-screenblock-inner">
                <h1>{% trans "This page has no preview!" %}</h1>
                <p>{% trans "It is being redirected to:" %} <a href="{{ cms_toolbar.redirect_url }}">{{ cms_toolbar.redirect_url }}</a></p>
            </div>
        </div>
    {% endif %}
```


[So I have fixed the previous failing tests and i have added the same tests using superuser](https://github.com/divio/django-cms/files/2479963/tests.py.txt)
Now i have these failing tests : 

- `test_unpublishedPage_302EmulationToFallbackLanguage_superuser`

- `test_publishedPage_302EmulationToFallbackLanguage_superuser`

In detail I have changed this statement :

```
user = self.get_staff_user_with_no_permissions()
user.user_permissions.add(Permission.objects.get(codename='change_page'))
```

with this one

```
user = self.get_superuser()
```

If my tests don't return any error, your patch will resolve my case, too.

Otherwise, as I can see in In my environment, superuser has a 200 response without that blocking message in overlay, that **probably** is not the correct behaviour. 

If you emulate, indeed, this behaviour using a superuser, and then publish the italian subpage with the wrong slug in the url (the english one), the system tries to let you see the current page published, but you will have an obvious 404 page because that link does not exists (it was taken from current `page_obj.title_cache`)
Unfortunately I'm not able to reproduce also this behavior with an automated test 
(maybe @toffi9 @czpython or @stefanfoulis could help me writing this test (please ignore me if you're busy) )

I hope I was clear in my explanation
Thank you in advance!

**Comment by mekanics** [comment]
@madEng84 

I think I got it. We just forgot the `PagePermission` itself. A user with `change_page` permission (Staff User) does not necessarily have the rights to change a page. He also need the 'can_change' permission on the page itself. Therefor we need to add this line of code: 

`PagePermission.objects.create(can_change=True, user=user, page=cms_page)` 

 A Superuser has all rights, so this is not needed. 

Now back to your tests. A Superuser or a Staff user with all the needed change permissions should be redirected to the appropriate page (`/it/contacts/work-with-us/`-> `/it/contatti/lavora-con-noi/`) with a status code 302 whether published or not. 

A user without the change permission should get a 200 with the overlay. 

I made some changes to your tests, see here: 
[test_madeng84.py.txt](https://github.com/divio/django-cms/files/2481844/test_madeng84.py.txt)

(I also had to change the used template and the language (it -> de), as italian was not in the `CMS_LANGUAGES` settings)

**Comment by madEng84** [comment]
perfect @mekaniks

The 302 was what i expected in my first test, if you remember.
I don't have a 302 response for a superuser in my local environment, i see two failures in my tests (200 response without overlay message).

So your patch fixes this issue! Great!

I hope this pull request is accepted in a short time.
If I can do something to help you I'm here.
Feel free to add these tests to your commit.

Thank you for your time!

**Comment by vThaian** [comment]
Thank you @mekanics and @madEng84 for your contribution. Could you please fix tests? Also, I would like to merge it to `develop` and then I will backport it to `3.6.X` and `3.5.X`. Thanks again :)

**Comment by FinalAngel** [comment]
@mekanics would you be able to adapt the PR according to the comments above?

**Comment by mekanics** [comment]
@vThaian @FinalAngel I'm not sure if the failing tests have anything to do with this fix, as they were failing on the source branch (release/3.5.x) before. Should I try to rebase the branch on the develop?

**Comment by FinalAngel** [comment]
that be amazing @mekanics

**Comment by mekanics** [comment]
can someone with enough right start the ci, please?

**Comment by mekanics** [comment]
@goutnet I just rebased my branch onto develop. The conflicts has been resolved. 

Now, we just need someone to review and merge the branch 🙂

**Comment by goutnet** [comment]
@FinalAngel  could I ask your wisedom on this maybe :) ?

**Comment by NicolaiRidani** [comment]
@mekanics  @goutnet You guys can also ask our dedicated PR review group on Slack #workgroup-pr-review to review the PR. See also our PR review process: https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst

File: CHANGELOG.rst [hunk 1]
```diff
@@ -23,9 +23,8 @@ Unreleased
 * Fixed missing builtin arguments on main ``cms`` management command causing it to crash
 * Fixed template label nested translation
 * Few changes in docs/contributing/code.rst
-* Add support for Django 3.2 LTS version
```
Comment (mekanics) [inline CHANGELOG.rst:26]: @goutnet Was this intended? 
Happened in your merge (develop into this branch)

Comment (goutnet) [reply CHANGELOG.rst:26]: hum, no shouldn't be there, can you either remove the merge and rebase, or revert the patch?

sorry for that

**Comment by goutnet** [comment]
@mekanics we seem to have a few broken tests, can you check?

**Review by goutnet — APPROVED** [review]
LGTM

**Comment by goutnet** [comment]
adding this to my bucket for 3.9.0

**Comment by goutnet** [comment]
I am restarting the tests to see how they goes, this one in next on rebase list :)

**Pull request merged**